### ISSUE TITLE
Adding a generic class attribute to att.common

### ISSUE THREAD

**Issue by lpugin** [music-encoding/music-encoding#338]
It could be interesting for many project to have a generic `@class` attribute for passing information from the MEI to processed output. I am thinking typically of SVG output. Its use would be completely project specific, similarly as with the `@type` attribute.

**Comment by pe-ro** [comment]
Sorry, but I don't see how `@class` is different from `@type`.  In nearly every case where `@type` is allowed, it can take any single-token value the user desires.  Is this not sufficient?

**Comment by kepper** [comment]
I could imagine it to be quite helpful when thinking about something like CSS for music, as we discussed in Montreal. You could add the same value to a group of dynams (for instance), and then express (elsewhere) that these things should be vertically aligned. Your explanation also points out a limitation: a single token. In HTML/@class, it's a series of tokens. I would prefer to keep it aligned with that (and maybe that's already an argument to have it separate from / in addition to type). I'm actually more thinking about the suggestion to have it completely free of any further specification / restriction / suggestions about potential use. If we plan to go in the direction of something like music:css later on, I wonder if a too loose definition right now may cause problems in the future – but I'm not sure about this yet…

**Comment by pe-ro** [comment]
The problem with "a series of tokens" is that to an XML parser it's indistinguishable from a multi-part, single value.  For instance, is `"this value"` 2 tokens or a single value?  Using a separator other than a space, often a semi-colon, as in `"this value; otherwise"`, clears up this problem, but requires additional parsing.  Of course, `@class` values could be defined as a series of single tokens, but it would be impossible to stop abuse without extraordinary measures.

**Comment by pe-ro** [comment]
If it's multiple tokens that are necessary, why not just re-define `@type`?  For instance,

```
 <attDef ident="type" usage="opt">
  <desc>Characterizes the element in some sense, using any convenient classification scheme or typology.</desc>
  <datatype maxOccurs="unbounded">
    <rng:data type="NMTOKEN"/>
  </datatype>
 </attDef>
```

**Comment by lpugin** [comment]
I suggested to add `@class` because `@type` does not allow multiple tokens and also because it is not available everywhere. Now if we change this two points then there is indeed maybe no need to add `@class`.

**Comment by kepper** [comment]
While the distinction is certainly not easy, I believe there is a significant difference between type and class. Type seems to be a generic (hierarchical) classification mechanism, while I understand class as a grouping mechanism: (in my understanding,) it allows to identify a set of elements, just like an xml:id identifies a single element. One could argue that this is equally possible with type (or even annot, if you like), but for me, it feels like twisting these concepts' arms to mimic the behavior of HTML's class attribute. If that's the case, I'd prefer to "steal" the original instead of overloading existing concepts. 

I'm not absolutely settled, though…

**Comment by craigsapp** [comment]
I would find `@class` useful.  A particular application that I can think of is to highlight notes, such as for search results.  For example here is a search for the pitch sequence "c d e f g" on the JRP website:

http://josquin.stanford.edu/data?a=notation&f=Agr1004c&themax=-D%22cdefg%22

The search results are highlighted in orange in this example.  The searching and labeling process are done symbolically, and then passed to the notation renderer.  This information is not for being stored in the data, but as a temporary pass-through to the rendered notation.  In other words, `@class` is particularly useful for doing analytic markup of existing scores.  

This searching would most conveniently be done by adding a `@class="match"` to the matched notes in the musical data, and then passing that attribute value into the SVG elements related to the MEI element.  Then CSS can be used to style the matched notes.   Note that this is not intended for a property of the score, but useful for post-processing of the score, such as searching in this case or for many other types of dynamic or analytic markup of the data found in the score.  Note that it would be inconvenient to do this as some sort of side-channel information either outside of the MEI file (how I currently am doing it with the Humdrum to MEI conversion), or internally, but based on a separate list of xml:ids.  As @kepper notes, `@class` is a grouping mechanism, which is convenient for grouping elements from different locations in the original tree structure of the data.  `@class` is most convenient as a pass-through from analytic markup in the MEI score, with the same behavior as in HTML: as a list of tokens separated by spaces (no multi-word tokens or non-space separation symbols).

Another practical example are the `@class=trill` stylings which I use to animate trills in SVG images.  `@class` is useful to transmit to the SVG notation what the function of a notational object is, since this information would otherwise be stripped out in the conversion to an image.  Having to refer back to the original MEI data via an xml:id in the SVG image would be cumbersome in most cases in terms of efficiency and storage: (1) searching for an ID takes time, and (2) some people may want to protect the MEI data from exposure to copying, and may prefer to not expose it in a final application, which would make back references to the original data not possible.  Here is a example SVG image which has the trill markup in measure 9:

http://www.ccarh.org/chopin/op24n2

![screen shot 2016-08-24 at 10 22 02 pm](https://cloud.githubusercontent.com/assets/3487289/17946584/39b6be9c-6a49-11e6-9658-04742f77a296.png)

The trilled note is wrapped in this <g> element in the SVG data:

<g class="noteon-25 noteoff-26 trill">
</g>

The trill class is identified from the symbolic notation (SCORE data in this case), and then a `@class="trill"`is applied to the SVG elements for that note.   Likewise, the class tags "noteon-25" and "noteoff-26" are used to keep track of the starting and ending times of the note (the note starts at quarter note time unit 25 in the score, and ends at quarter note time unit 26).   All of this information is used to animate the SVG notation with recording playback without needing to refer to the original score since all necessary information is not present in the SVG image for playback.  In this case the duration of the note is one quarter note (26-25=1), and the duration of the note can be calculated from the tempo (of the real performance in this example), and an animation cycle count for the trill is given by that duration and how fast the animation should cycle.

For the searching and trill animation examples, how would these be done other than with `@class`, and would another method be better?

**Comment by pe-ro** [comment]
@kepper : I've tried splitting this particular hair before, although not in quite the same way as you.  My experience with that leads me to believe that fine distinctions like yours don't work in practice.  In my opinion, `@type` is simply a generic, _ad hoc_ extension and restriction mechanism, whether it's applied to a single element or to many.  Of course, when it's applied to several elements, then those elements are members of a group that shares the given extension or restriction.

In the following example, each measure is deemed to be either "good" or "bad" according to some definition known only to the encoder.  Since measures 1 & 3 are both "good", they belong to the group/type/class "good".

```
<measure type="good"/>
<measure type="bad"/>
<measure type="good"/>
```

Allowing `@type` to contain multiple tokens, as long as those tokens don't contain any space characters, will permit the "layering" of multiple extension/restriction systems.  For example, an element can be both "fish" and "fowl" simultaneously and values assigned by one encoder don't have to be replaced by those of another encoder:

```
<measure type="fish fowl good bad"/>
```

There's no mechanism, however, for resolving conflicts like the one exemplified here.

Currently, `@subtype` also only permits a single value.  We can change that, but when `@type` is multi-valued, there's no automated way to match up type and subtype values:

```
<measure type="good bad" subtype="enough tothebone "/>
```

One could rely on encoding order to figure out that "enough" goes with "good" and "tothebone" goes with "bad", but that may not always the case.  A better approach is to use combined type-and-subtype values, such as "good-to-the-bone" and "bad-enough".

At a certain level of complexity, one may rely on the `<classification>` element and its children, which is currently accessed using `@decls`:

```
<measure decls="#term1 #term2"/>
```

`<classification>` allows the use of standard or user-created classification/grouping mechanisms, i.e., thesauri and ontologies.  Since this is a specific usage of `@decls`, which is a very generic way of pointing at something/anything in the header, not just `<classification>`, I've been hoping to reserve `@class` for this as it makes a nice symmetry, i.e., `@class` always points to `<classification>`.

To summarize, I'd prefer to:
- allow `@type` and `@subtype` to take multiple, single-token values,
- make `@type` and `@subtype` available globally, and
- add `@class` (also globally) as a way of linking elements to `<classification>` for complex cases.

**Comment by lpugin** [comment]
I forgot we had `@subtype` too. For me, this makes allowing `@type` to take multiple values not a great solution. Even though they are generic independent attributes, they undoubtedly imply some sort of hierarchical relation (subtype of a type). With multiple values allowed, we loose this and `@subtype` loses his sense. Further more, changing this makes us diverge from TEI practice, which will introduce additional confusion. So sorry for changing my mind, but it now seems to me that adding `@class` globally with multiple, single-token values and leaving `@type` untouched might be better.

**Comment by pe-ro** [comment]
@lpugin : In my opinion, this is a recipe for confusion between `@type` and `@class`.  A better solution is to allow `@type` to take multiple values and simply remove `@subtype`.  This option is better because it's less disruptive for existing MEI instances.  If `@subtype` is in use, its value can be combined with that of `@type`.  For example, `type="foo" subtype="bar"` can be replaced by `type="foo-bar"`.  I'm not particularly concerned about compliance with TEI practice.  Attributes that depend on other attributes is generally a bad idea anyway, so I'm not eager to follow them over that particular cliff.

**Comment by pe-ro** [comment]
I forgot to say that this proposal allows any elements that currently have a list of suggested values for `@type` to retain them. So, 

```
<div type="abstract something-else thisToo"/>
```

is fine.

**Comment by ahankinson** [comment]
Could `type` vs. `class` be clarified with the distinction between `is-a` and `belongs-to`? I think of type being something descriptive about an object, whereas classes are groupings of like objects. I think of classes defining an ad-hoc grouping of elements (a bit like the opt-in pattern for `classSpec` in ODD.)

I'm not so big on replicating `@class` in the HTML/SVG sense, but as an ad-hoc grouping mechanism it may be useful.

**Comment by pe-ro** [comment]
Like I attempted to say before, the distinction between `is-a` and `belongs-to` may become fuzzy in the heat of encoding, so I'd like to avoid setting up the choice.  For instance, something that `is` red can be thought of as `belonging to` a class of things that share that quality.  So, I prefer a solution that provides for both interpretations.  I agree that replicating HTML's `@class` attribute isn't a priority, but that's where this is headed.

**Comment by lpugin** [comment]
> A better solution is to allow @type to take multiple values and simply remove @subtype. This option is better because it's less disruptive for existing MEI instances. 

Removing an existing attribute is clearly more disruptive than adding a new one.

**Comment by pe-ro** [comment]
But I believe no one is actually using `@subtype`.

**Comment by kepper** [comment]
I've used it once or twice during the Freischütz, but only for temporary things. I felt pretty alone doing so, and I wonder if anyone else ever used it. And actually, we could have done it equally well with multiple values in type. If we go into that direction, I would like to see a clear suggestion on how subtypes should be converted into additional types (personally, I'd prefer dots as "hierarchy indicators" like "type.subtype", but that's just a matter of preferred programming languages and as such doesn't really make a difference…). With such a clear statement, we can make use of that in the XSLT that "upconverts" into the next release, and it would also be possible to "downconvert" newer instances. But actually, killing subtype in favor of additional type values gives projects the freedom to have any level of "type-nesting", not just two (but that would make "downconversion" harder). In essence, I start to agree with Perry.

Just my current 2c,
jo

**Comment by lpugin** [comment]
Any objection to dropping `@subtype` and making `@type` available globally and taking multiple token?

**Comment by pe-ro** [comment]
@music-encoding/core-contributors ?

**Comment by ahankinson** [comment]
Sounds ok with me.

**Comment by kepper** [comment]
not sure if I qualify as core contributor, but I'm all in.

**Comment by pe-ro** [comment]
Unless we hear otherwise in the next couple of days, I think we can move forward with this.

**Comment by raffazizzi** [comment]
Sorry to jump in a little late. I don't disagree with the conclusion of this discussion (make `@type` take multiple values and deprecate `@subtype`). My only itch is that this breaks a parallel with TEI. But, ultimately, what's the point of keeping that parallel?
So I'd say go ahead.

**Comment by kepper** [comment]
and maybe we can serve as model for TEI? ;-) Honestly, I believe the restriction to a two-level hierarchy, with only single values on both levels, is actually quite severe. TEI surely can't change that as easily as MEI, but with a clear suggestion how to convert existing data (for instance, @type="level1" @subtype="level2" could translate to @type="level1 level1_level2") I believe it's not undoable. Considering how much flexibility can be gained by allowing multiple values for @type, I would at least think about it. However, it's up to you to bring that idea to TEI's table ;-)

**Comment by raffazizzi** [comment]
I think I will. As you said it's unlikely that they'd want to break backward compatibility, but I'm also curious to hear the rest of the council's arguments for the two-level hierarchy. FWIW I often bring up useful MEI practices; what happens here has provided a very helpful model for migrating TEI to GitHub, for example.

**Comment by pe-ro** [comment]
@raffazizzi: Any news from the TEI front that might dissuade us from moving ahead on this issue?
